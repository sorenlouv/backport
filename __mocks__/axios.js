const axios = jest.genMockFromModule('axios');
axios.get = jest.fn(() => {
  throw new Error('Forgot to mock axios.get');
});
axios.post = jest.fn(() => {
  throw new Error('Forgot to mock axios.post');
});
module.exports = axios;
