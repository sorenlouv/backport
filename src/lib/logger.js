const logger = {};
logger.log = (...args) => {
  console.log(...args);
};

logger.error = (...args) => {
  console.error(...args);
};

module.exports = logger;
