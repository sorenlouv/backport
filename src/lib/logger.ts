export function log(...args: unknown[]) {
  console.log(...args);
}

export function error(...args: unknown[]) {
  console.error(...args);
}
