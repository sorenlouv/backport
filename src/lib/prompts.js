const inquirer = require('inquirer');

function prompt(options) {
  return inquirer
    .prompt([Object.assign({}, options, { name: 'promptResult' })])
    .then(({ promptResult }) => promptResult);
}

function listProjects(repoNames) {
  return prompt({
    type: 'list',
    message: 'Select project',
    choices: repoNames
  });
}

function listCommits(commits, isMultipleChoice) {
  const pageSize = Math.min(10, commits.length);
  return prompt({
    pageSize,
    type: isMultipleChoice ? 'checkbox' : 'list',
    message: 'Select commit to backport',
    choices: commits
      .map(commit => ({
        name: commit.message,
        value: commit,
        short: commit.message
      }))
      .concat(commits.length > pageSize ? new inquirer.Separator() : [])
  })
    .then(commit => (isMultipleChoice ? commit.reverse() : [commit]))
    .then(
      selectedCommits =>
        selectedCommits.length === 0
          ? listCommits(commits, isMultipleChoice)
          : selectedCommits
    );
}

function listBranches(branches, isMultipleChoice) {
  return prompt({
    type: isMultipleChoice ? 'checkbox' : 'list',
    message: 'Select branch to backport to',
    choices: branches
  })
    .then(res => (isMultipleChoice ? res : [res]))
    .then(
      selectedBranches =>
        selectedBranches.length === 0
          ? listBranches(branches, isMultipleChoice)
          : selectedBranches
    );
}

function confirmConflictResolved() {
  return prompt({
    type: 'confirm',
    message: 'Press enter when you have commited all changes'
  });
}

module.exports = {
  confirmConflictResolved,
  listCommits,
  listProjects,
  listBranches
};
