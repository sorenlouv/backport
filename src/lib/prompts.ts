import inquirer, { Question } from 'inquirer';
import isEmpty from 'lodash.isempty';
import { Commit, BranchChoice } from '../types/types';

async function prompt<T>(options: Question) {
  const { promptResult } = (await inquirer.prompt([
    { ...options, name: 'promptResult' }
  ])) as { promptResult: T };
  return promptResult;
}

export function listProjects(repoNames: string[]) {
  return prompt({
    choices: repoNames,
    message: 'Select project',
    type: 'list'
  });
}

export async function listCommits(
  commits: Commit[],
  isMultipleChoice: boolean
): Promise<Commit[]> {
  const choices = commits.map((c, i) => ({
    name: `${i + 1}. ${c.message}`,
    short: c.message,
    value: c
  }));

  const res = await prompt({
    choices,
    message: 'Select commit to backport',
    pageSize: Math.min(10, commits.length),
    type: isMultipleChoice ? 'checkbox' : 'list'
  });

  const selectedCommits = Array.isArray(res) ? res.reverse() : [res];
  return isEmpty(selectedCommits)
    ? listCommits(commits, isMultipleChoice)
    : selectedCommits;
}

export async function listBranches(
  branches: BranchChoice[],
  isMultipleChoice: boolean
): Promise<string[]> {
  const res = await prompt<string | string[]>({
    choices: branches,
    message: 'Select branch to backport to',
    type: isMultipleChoice ? 'checkbox' : 'list'
  });

  const selectedBranches = Array.isArray(res) ? res : [res];

  return isEmpty(selectedBranches)
    ? listBranches(branches, isMultipleChoice)
    : selectedBranches;
}

export function confirmConflictResolved() {
  return prompt({
    message: 'Press enter when you have commited all changes',
    type: 'confirm'
  });
}
