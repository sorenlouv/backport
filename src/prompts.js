const inquirer = require('inquirer');

function prompt(options) {
  return inquirer
    .prompt([Object.assign({}, options, { name: 'promptResult' })])
    .then(({ promptResult }) => promptResult);
}

function listFullRepoName(repoNames) {
  return prompt({
    type: 'list',
    message: 'Select repository',
    choices: repoNames
  });
}

function listCommits(commits) {
  const pageSize = Math.min(10, commits.length);
  return prompt({
    pageSize,
    type: 'list',
    message: 'Select commit to backport',
    choices: commits
      .map(commit => ({
        name: commit.message,
        value: commit,
        short: commit.message
      }))
      .concat(commits.length > pageSize ? new inquirer.Separator() : [])
  });
}

function listVersions(versions) {
  return prompt({
    type: 'list',
    message: 'Select version to backport to',
    choices: versions
  }).then(version => [version]);
}

function checkboxVersions(versions) {
  return prompt({
    type: 'checkbox',
    message: 'Select version to backport to',
    choices: versions
  });
}

function confirmConflictResolved() {
  return prompt({
    type: 'confirm',
    message: 'Have you solved the merge conflict?'
  });
}

module.exports = {
  confirmConflictResolved,
  listCommits,
  listFullRepoName,
  listVersions,
  checkboxVersions
};
