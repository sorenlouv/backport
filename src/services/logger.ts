import winston, { format } from 'winston';
import yargs from 'yargs';
import { getLogfilePath } from './env';
import safeJsonStringify from 'safe-json-stringify';
import isString from 'lodash.isstring';
import dedent from 'dedent';

const { combine } = format;
const { argv } = yargs.help(false);

// wrapper around console.log
export function consoleLog(...args: unknown[]) {
  console.log(...args);
}

const level = argv.verbose ? 'verbose' : argv.debug ? 'debug' : 'info';

let winstonInstance: winston.Logger;

export const logger = {
  info: (message: string, meta?: unknown) => {
    if (winstonInstance) {
      winstonInstance.info(message, { meta });
    }
  },
  verbose: (message: string, meta?: unknown) => {
    if (winstonInstance) {
      winstonInstance.verbose(message, { meta });
    }
  },
  debug: (message: string, meta?: unknown) => {
    if (winstonInstance) {
      winstonInstance.debug(message, { meta });
    }
  },
};

export function initLogger() {
  winstonInstance = winston.createLogger({
    transports: [
      // log to file
      new winston.transports.File({
        level,
        format: combine(
          format.timestamp({ format: 'YYYY-MM-DD HH:mm:ss' }),
          winston.format.metadata({
            fillExcept: ['message', 'level', 'timestamp', 'label'],
          }),

          format.printf((info) => {
            if (!info.metadata.meta) {
              return `${info.timestamp}: ${info.message}`;
            }

            if (isString(info.metadata.meta)) {
              return `${info.timestamp} ${info.message}\n${dedent(
                info.metadata.meta
              )}\n`;
            }

            return `${info.timestamp} ${info.message}\n${safeJsonStringify(
              info.metadata.meta,
              null,
              2
            )}\n`;
          })
        ),
        filename: getLogfilePath(),
      }),
    ],
  });
  return winstonInstance;
}

// log levels:
// - error
// - warn
// - info
// - verbose
// - debug
