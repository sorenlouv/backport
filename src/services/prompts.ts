import chalk from 'chalk';
import inquirer, {
  CheckboxQuestion,
  ListQuestion,
  ConfirmQuestion,
} from 'inquirer';
import isEmpty from 'lodash.isempty';
import { TargetBranchChoice } from '../options/ConfigOptions';
import { Commit } from '../types/Commit';
import { getShortSha } from './github/commitFormatters';

type Question = CheckboxQuestion | ListQuestion | ConfirmQuestion;

async function prompt<T = unknown>(options: Question) {
  const { promptResult } = (await inquirer.prompt([
    { ...options, name: 'promptResult' },
  ])) as { promptResult: T };
  return promptResult;
}

export async function promptForCommits({
  commitChoices,
  isMultipleChoice,
}: {
  commitChoices: Commit[];
  isMultipleChoice: boolean;
}): Promise<Commit[]> {
  const choices = commitChoices.map((c, i) => {
    const existingPRs = c.existingTargetPullRequests
      .map((item) => {
        const styling = item.state === 'MERGED' ? chalk.green : chalk.gray;
        return styling(item.branch);
      })
      .join(', ');

    const position = chalk.gray(`${i + 1}.`);

    return {
      name: `${position} ${c.formattedMessage} ${existingPRs}`,
      short: c.pullNumber
        ? `#${c.pullNumber} (${getShortSha(c.sha)})`
        : getShortSha(c.sha),
      value: c,
    };
  });

  const res = await prompt<Commit[]>({
    loop: false,
    pageSize: 15,
    choices: choices,
    message: 'Select commit',
    type: isMultipleChoice ? 'checkbox' : 'list',
  });

  const selectedCommits = Array.isArray(res) ? res.reverse() : [res];
  return isEmpty(selectedCommits)
    ? promptForCommits({ commitChoices: commitChoices, isMultipleChoice })
    : selectedCommits;
}

export async function promptForTargetBranches({
  targetBranchChoices,
  isMultipleChoice,
}: {
  targetBranchChoices: TargetBranchChoice[];
  isMultipleChoice: boolean;
}): Promise<string[]> {
  const res = await prompt<string | string[]>({
    loop: false,
    pageSize: 15,
    choices: targetBranchChoices,
    message: 'Select branch',
    type: isMultipleChoice ? 'checkbox' : 'list',
  });

  const selectedBranches = Array.isArray(res) ? res : [res];

  return isEmpty(selectedBranches)
    ? promptForTargetBranches({
        targetBranchChoices,
        isMultipleChoice,
      })
    : selectedBranches;
}

export function confirmPrompt(message: string) {
  return prompt<boolean>({ message, type: 'confirm' });
}
