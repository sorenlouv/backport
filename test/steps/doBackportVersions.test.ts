import axios from 'axios';
import * as childProcess from 'child_process';
import {
  doBackportVersion,
  getReferenceLong
} from '../../src/steps/doBackportVersions';
import { PromiseReturnType } from '../../src/types/commons';
import last from 'lodash.last';
import * as prompts from '../../src/services/prompts';
import * as logger from '../../src/services/logger';

describe('doBackportVersion', () => {
  let axiosMockInstance: jest.Mock;

  beforeEach(() => {
    axiosMockInstance = jest
      .spyOn(axios, 'post')
      // mock: createPullRequest
      // @ts-ignore
      .mockResolvedValueOnce({
        data: {
          number: 1337,
          html_url: 'myHtmlUrl'
        }
      })
      // mock: addLabelsToPullRequest
      .mockResolvedValueOnce(null);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  describe('when commit has a pull request reference', () => {
    let execSpy: jest.SpyInstance;
    let res: PromiseReturnType<typeof doBackportVersion>;
    beforeEach(async () => {
      const commits = [
        {
          sha: 'mySha',
          message: 'myCommitMessage',
          pullNumber: 1000
        },
        {
          sha: 'mySha2',
          message: 'myOtherCommitMessage',
          pullNumber: 2000
        }
      ];

      execSpy = jest.spyOn(childProcess, 'exec');

      res = await doBackportVersion(
        'elastic',
        'kibana',
        commits,
        '6.x',
        'sqren',
        ['backport'],
        '[{baseBranch}] {commitMessages}',
        'myPrSuffix',
        'api.github.com'
      );
    });

    it('should make correct git commands', () => {
      expect(execSpy.mock.calls).toMatchSnapshot();
    });

    it('should return correct response', () => {
      expect(res).toEqual({ html_url: 'myHtmlUrl', number: 1337 });
    });

    it('should create pull request and add labels', () => {
      expect(axiosMockInstance).toHaveBeenCalledTimes(2);
      expect(axiosMockInstance).toHaveBeenNthCalledWith(
        1,
        'https://api.github.com/repos/elastic/kibana/pulls?access_token=undefined',
        {
          title: '[6.x] myCommitMessage | myOtherCommitMessage',
          body:
            'Backports the following commits to 6.x:\n - myCommitMessage (#1000)\n - myOtherCommitMessage (#2000)\n\nmyPrSuffix',
          head: 'sqren:backport/6.x/pr-1000_pr-2000',
          base: '6.x'
        }
      );

      expect(axiosMockInstance).toHaveBeenNthCalledWith(
        2,
        'https://api.github.com/repos/elastic/kibana/issues/1337/labels?access_token=undefined',
        ['backport']
      );
    });
  });

  describe('when commit does not have a pull request reference', () => {
    beforeEach(async () => {
      const commits = [
        {
          sha: 'mySha',
          message: 'myCommitMessage'
        }
      ];

      await doBackportVersion(
        'elastic',
        'kibana',
        commits,
        '6.x',
        'sqren',
        ['backport'],
        '[{baseBranch}] {commitMessages}',
        undefined,
        'api.github.com'
      );
    });

    it('should create pull request and add labels', () => {
      expect(axiosMockInstance).toHaveBeenCalledTimes(2);
      expect(axiosMockInstance).toHaveBeenNthCalledWith(
        1,
        'https://api.github.com/repos/elastic/kibana/pulls?access_token=undefined',
        {
          title: '[6.x] myCommitMessage',
          body:
            'Backports the following commits to 6.x:\n - myCommitMessage (mySha)',
          head: 'sqren:backport/6.x/commit-mySha',
          base: '6.x'
        }
      );

      expect(axiosMockInstance).toHaveBeenNthCalledWith(
        2,
        'https://api.github.com/repos/elastic/kibana/issues/1337/labels?access_token=undefined',
        ['backport']
      );
    });
  });

  describe('when cherry-picking fails', () => {
    function didResolveConflict(didResolve: boolean) {
      const logSpy = jest.spyOn(logger, 'log');
      const commits = [
        {
          sha: 'mySha',
          message: 'myCommitMessage'
        }
      ];

      const execSpy = jest
        .spyOn(childProcess, 'exec')
        .mockImplementation((...args: any[]) => {
          const [cmd] = args;
          if (cmd.includes('git cherry-pick')) {
            const e = new Error('as');
            // @ts-ignore
            e.cmd = cmd;
            throw e;
          } else {
            last(args)();
          }

          return {} as any;
        });

      spyOn(prompts, 'confirmPrompt').and.returnValue(didResolve);

      const promise = doBackportVersion(
        'elastic',
        'kibana',
        commits,
        '6.x',
        'sqren',
        ['backport'],
        'myPrTitle',
        undefined,
        'api.github.com'
      );

      return { logSpy, execSpy, promise };
    }

    it('and conflicts were resolved', async () => {
      const { execSpy, promise } = didResolveConflict(true);
      await promise;
      expect(execSpy.mock.calls).toMatchSnapshot();
      expect(axiosMockInstance).toHaveBeenCalledTimes(2);
    });

    it('and conflicts were not resolved', async () => {
      const { execSpy, promise, logSpy } = didResolveConflict(false);
      expect.assertions(4);

      await promise.catch(e => {
        expect(logSpy.mock.calls).toMatchInlineSnapshot(`
          Array [
            Array [
              "Backporting mySha to 6.x:",
            ],
            Array [
              "Please resolve conflicts in: /myHomeDir/.backport/repositories/elastic/kibana and when all conflicts have been resolved and staged run:",
            ],
            Array [
              "
              git cherry-pick --continue
              ",
            ],
          ]
        `);
        expect(e.message).toEqual('Aborted');
        expect(execSpy.mock.calls).toMatchSnapshot();
        expect(axiosMockInstance).toHaveBeenCalledTimes(0);
      });
    });
  });
});

describe('getReferenceLong', () => {
  it('should return a sha', () => {
    expect(
      getReferenceLong({ sha: 'mySha1234567', message: 'myMessage' })
    ).toEqual('mySha12');
  });

  it('should return a pr', () => {
    expect(
      getReferenceLong({
        pullNumber: 1337,
        sha: 'mySha1234567',
        message: 'myMessage'
      })
    ).toEqual('#1337');
  });
});
