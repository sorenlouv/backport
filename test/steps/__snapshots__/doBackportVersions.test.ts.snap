// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`doBackportVersion when cherry-picking fails and conflicts were not resolved 2`] = `
Array [
  Array [
    "git reset --hard && git clean -d --force && git checkout master && git pull origin master",
    Object {
      "cwd": "/myHomeDir/.backport/repositories/elastic/kibana",
      "maxBuffer": 104857600,
    },
    [Function],
  ],
  Array [
    "git fetch origin 6.x && git branch backport/6.x/commit-mySha origin/6.x --force && git checkout backport/6.x/commit-mySha ",
    Object {
      "cwd": "/myHomeDir/.backport/repositories/elastic/kibana",
      "maxBuffer": 104857600,
    },
    [Function],
  ],
  Array [
    "git cherry-pick mySha",
    Object {
      "cwd": "/myHomeDir/.backport/repositories/elastic/kibana",
      "maxBuffer": 104857600,
    },
    [Function],
  ],
]
`;

exports[`doBackportVersion when cherry-picking fails and conflicts were resolved 1`] = `
Array [
  Array [
    "git reset --hard && git clean -d --force && git checkout master && git pull origin master",
    Object {
      "cwd": "/myHomeDir/.backport/repositories/elastic/kibana",
      "maxBuffer": 104857600,
    },
    [Function],
  ],
  Array [
    "git fetch origin 6.x && git branch backport/6.x/commit-mySha origin/6.x --force && git checkout backport/6.x/commit-mySha ",
    Object {
      "cwd": "/myHomeDir/.backport/repositories/elastic/kibana",
      "maxBuffer": 104857600,
    },
    [Function],
  ],
  Array [
    "git cherry-pick mySha",
    Object {
      "cwd": "/myHomeDir/.backport/repositories/elastic/kibana",
      "maxBuffer": 104857600,
    },
    [Function],
  ],
  Array [
    "git diff-index --quiet HEAD --",
    Object {
      "cwd": "/myHomeDir/.backport/repositories/elastic/kibana",
      "maxBuffer": 104857600,
    },
    [Function],
  ],
  Array [
    "git push sqren backport/6.x/commit-mySha:backport/6.x/commit-mySha --force",
    Object {
      "cwd": "/myHomeDir/.backport/repositories/elastic/kibana",
      "maxBuffer": 104857600,
    },
    [Function],
  ],
]
`;

exports[`doBackportVersion when commit has a pull request reference should make correct git commands 1`] = `
Array [
  Array [
    "git reset --hard && git clean -d --force && git checkout master && git pull origin master",
    Object {
      "cwd": "/myHomeDir/.backport/repositories/elastic/kibana",
      "maxBuffer": 104857600,
    },
    [Function],
  ],
  Array [
    "git fetch origin 6.x && git branch backport/6.x/pr-1000_pr-2000 origin/6.x --force && git checkout backport/6.x/pr-1000_pr-2000 ",
    Object {
      "cwd": "/myHomeDir/.backport/repositories/elastic/kibana",
      "maxBuffer": 104857600,
    },
    [Function],
  ],
  Array [
    "git cherry-pick mySha",
    Object {
      "cwd": "/myHomeDir/.backport/repositories/elastic/kibana",
      "maxBuffer": 104857600,
    },
    [Function],
  ],
  Array [
    "git cherry-pick mySha2",
    Object {
      "cwd": "/myHomeDir/.backport/repositories/elastic/kibana",
      "maxBuffer": 104857600,
    },
    [Function],
  ],
  Array [
    "git push sqren backport/6.x/pr-1000_pr-2000:backport/6.x/pr-1000_pr-2000 --force",
    Object {
      "cwd": "/myHomeDir/.backport/repositories/elastic/kibana",
      "maxBuffer": 104857600,
    },
    [Function],
  ],
]
`;
